src/WebJobs.Extensions.Http/ClaimsPrincipalBindingProvider.cs

```diff
@@ -0,0 +1,119 @@
+
+            public ClaimsIdentityBinding(ParameterInfo parameter)
+            {
+                _parameter = parameter;
+            }
+
+            public bool FromAttribute
+            {
+                get { return false; }
+            }
+
+            public Task<IValueProvider> BindAsync(BindingContext context)
+            {
+                if (context == null)
+                {
+                    throw new ArgumentNullException("context");
+                }
+
+                IReadOnlyDictionary<string, object> bindingData = context.BindingData;
+                ClaimsPrincipal principal = GetClaimsPrincipalFromBindingData(bindingData);
+
+                var valueProvider = new SimpleValueProvider(typeof(ClaimsPrincipal), principal, principal?.Identity?.Name);
+                return Task.FromResult<IValueProvider>(valueProvider);
+            }
+
+            private static ClaimsPrincipal GetClaimsPrincipalFromBindingData(IReadOnlyDictionary<string, object> bindingData)
+            {
+                object headers;
+                if (bindingData.TryGetValue(HttpTriggerAttributeBindingProvider.HttpHeadersKey, out headers))
+                {
+                    var requestHeaders = new Dictionary<string, string>(headers as IDictionary<string, string>, StringComparer.OrdinalIgnoreCase);
+                    List<ClaimsIdentity> identities = PotentialIdentityHeaders
+                        .Where(potentialHeader => requestHeaders.ContainsKey(potentialHeader))
+                        .Select(headerKey => ClaimsIdentityHelper.FromBase64EncodedJson(requestHeaders[headerKey]))
+                        .ToList();
+                    if (identities.Count == 0)
+                    {
+                        // According to the source https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/security/principal/genericidentity.cs
+                        // GenericIdentity is unauthenticated if and only if the name is set to the empty string
+                        var unauthentiatedIdentity = new GenericIdentity(name: string.Empty);
+                        identities.Add(unauthentiatedIdentity);
+                    }
+                    return new ClaimsPrincipal(identities);
+                }
+                else
+                {
+                    return new ClaimsPrincipal();
+                }
+            }
+
+            public Task<IValueProvider> BindAsync(object value, ValueBindingContext context)
+            {
+                var request = value as ClaimsPrincipal;
+                if (request != null)
+                {
+                    var binding = new SimpleValueProvider(typeof(ClaimsPrincipal), request, "request");
+                    return Task.FromResult<IValueProvider>(binding);
+                }
+                throw new NotSupportedException($"A ClaimsPrincipal is required to bind using {nameof(ClaimsPrincipalBindingProvider)}.");
```

Comment:
>ClaimsPrincipalBindingProvider [](start = 102, length = 30)

Will the user now what a `ClaimsPrincipalBindingProvider` is? I assume not, in which case we'll want to describe the problem differently. Could we not say "A ClaimsPrincipal parameter type is required when binding to the remote user's identity" or something similar?